src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/Debugger/WebDebugTargetsProvider.cs

```diff
@@ -0,0 +1,431 @@
+using Microsoft.Win32;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Web
+{
+    /// <summary>
+    /// Provides QueryDebugTargetsAsync() support for web profile iisexpress, web command, IIS. It is 
+    /// not an exported CPS debugger but is imported by ProjectDebuggerProvider to handle IIS
+    /// and IIS Express web servers.
+    /// </summary>
+    [Export(typeof(IDebugProfileLaunchTargetsProvider))]
+    [AppliesTo(ProjectCapability.AspNetLaunchProfiles)]
+    [Order(100)]
+    internal class WebDebugTargetsProvider : IDebugProfileLaunchTargetsProvider, IDebugProfileLaunchTargetsProvider2, IDeployedProjectItemMappingProvider, IDisposable
+    {
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IDebugTokenReplacer _tokenReplacer;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IFileSystem _fileSystem;
+        private readonly WebServer _webServer;
+        private readonly IDebugTargetsObserver _debugTargetsObserver;
+        private readonly IVsUIService<SVsShell, IVsShell> _vsShell;
+        private readonly IVsUIService<SVsUIShellOpenDocument, IVsUIShellOpenDocument> _vsShellOpenDoc;
+        private readonly IVsUIService<IBrowserDebugTargetSelectionService, IBrowserDebugTargetSelectionService> _browserDebugTargetSelectionSvc;
+
+        [ImportingConstructor]
+        public WebDebugTargetsProvider(ConfiguredProject configuredProject,
+                                       IDebugTokenReplacer tokenReplacer,
+                                       IUnconfiguredProjectVsServices projectVsServices,
+                                       IProjectThreadingService threadingService,
+                                       IFileSystem fileSystem,
+                                       WebServer webServer,
+                                       IDebugTargetsObserver debugTargetsObserver,
+                                       IVsUIService<SVsShell, IVsShell> vsShell,
+                                       IVsUIService<SVsUIShellOpenDocument, IVsUIShellOpenDocument> vsShellOpenDoc,
+                                       IVsUIService<IBrowserDebugTargetSelectionService, IBrowserDebugTargetSelectionService> browserDebugTargetSelectionSvc)
+        {
+            _configuredProject = configuredProject;
+            _tokenReplacer = tokenReplacer;
+            _projectVsServices = projectVsServices;
+            _threadingService = threadingService;
+            _fileSystem = fileSystem;
+            _webServer = webServer;
+            _debugTargetsObserver = debugTargetsObserver;
+            _vsShell = vsShell;
+            _vsShellOpenDoc = vsShellOpenDoc;
+            _browserDebugTargetSelectionSvc = browserDebugTargetSelectionSvc;
+        }
+
+        public int ProcessIDToResume { get; protected set; }
```

Comment:
>public [](start = 8, length = 6)

should these be internal until there is code that uses them?

Reply:
I removed the unused ones, the rest I made private.